resource/lvm/fs/fs.go

```diff
@@ -0,0 +1,177 @@
+	}
+
+	{
+		fs, err := r.checkBlkid(r.mount.What)
+		if err != nil {
+			return nil, err
+		}
+		log.Debugf("blkid detect following fstype: %s, planned fstype: %s", fs, r.mount.Type)
+		if fs == r.mount.Type {
+			r.needMkfs = false
+		} else if fs == "" {
+			r.needMkfs = true
+			status.AddDifference("format", fs, r.mount.Type, "")
+		} else {
+			return nil, fmt.Errorf("%s already contain other filesystem with different type %s", r.mount.What, fs)
+		}
+	}
+
+	{
+		ok, err := r.lvm.CheckUnit(r.unitFileName, r.unitFileContent)
+		if err != nil {
+			return nil, err
+		}
+		if ok {
+			r.unitNeedUpdate = true
+			status.AddDifference(r.unitFileName, "<none>", r.unitFileContent, "")
+		}
+	}
+
+	// FIXME: check what device mounted to Where
+	{
+		ok, err := r.lvm.Mountpoint(r.mount.Where)
+		if err != nil {
+			return nil, err
+		}
+		r.mountNeedUpdate = r.unitNeedUpdate && !ok
+	}
+
+	if r.mountNeedUpdate {
+		status.AddDifference(r.mount.Where, "<none>", fmt.Sprintf("mount %s", r.mount.Where), "")
+	}
+
+	if resource.AnyChanges(status.Differences) {
+		status.Level = resource.StatusWillChange
+	}
+
+	return status, nil
+}
+
+func (r *resourceFS) Apply() (resource.TaskStatus, error) {
+	if r.needMkfs {
+		if err := r.lvm.Mkfs(r.mount.What, r.mount.Type); err != nil {
+			return nil, err
+		}
+	}
+	if r.unitNeedUpdate {
+		if err := r.lvm.UpdateUnit(r.unitFileName, r.unitFileContent); err != nil {
+			return nil, err
+		}
+	}
+
+	// FIXME: need mkdir
+	if r.mountNeedUpdate {
+		if err := r.lvm.StartUnit(r.unitServiceName()); err != nil {
+			return nil, err
+		}
+	}
+
+	return &resource.Status{}, nil
+}
+
+// FIXME: ugly kludge
+func (r *resourceFS) checkBlkid(name string) (string, error) {
+	if _, err := os.Stat(name); os.IsNotExist(err) {
+		return "", nil
+	}
+	return r.lvm.Blkid(name)
+}
+
+// NewResourceFS create new resource.Task node for create/mount FileSystem.
+func NewResourceFS(lvm lowlevel.LVM, m *Mount) (resource.Task, error) {
+	var err error
+	r := &resourceFS{
+		lvm:   lvm,
+		mount: m,
+	}
+	r.unitFileName = r.unitName()
+	r.unitFileContent, err = r.renderUnitFile()
+	if err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+func (r *resourceFS) escapedMountpoint() string {
+	// FIXME: proper systemd' escaping of r.mountpoint should be
+	return strings.Replace(strings.Trim(r.mount.Where, "/"), "/", "-", -1)
+}
+
+func (r *resourceFS) unitServiceName() string {
+	return fmt.Sprintf("%s.mount", r.escapedMountpoint())
+}
+
+func (r *resourceFS) unitName() string {
+	return fmt.Sprintf("/etc/systemd/system/%s", r.unitServiceName())
+}
+
+func (r *resourceFS) renderUnitFile() (string, error) {
+	var b bytes.Buffer
+	tmpl, err := template.New("unit.mount").Parse(unitTemplate)
+	if err != nil {
+		return "", err
+	}
+	err = tmpl.Execute(&b, r.mount)
+	if err != nil {
+		return "", err
+	}
+	return b.String(), nil
+}
+
+func init() {
+	registry.Register("lvm.fs", (*Preparer)(nil), (*resourceFS)(nil))
```

Comment:
It could be nice to call the modules something more human readable.  I know we're naming them after the underlying LVM commands and maybe that's best, but as someone who isn't familiar with LVM it would have been a lot quicker to understand what was happening if these were caled "lvm.filesystem", "lvm.volumegroup" and "lvm.physicalvolume".  Again I'm open to disagreement on this :-)